modules/internal/OwnedObject.chpl

```diff
@@ -245,16 +245,37 @@ module OwnedObject {
        refer to `nil` after this call.
      */
     proc init=(pragma "leaves arg nil" pragma "nil from arg" ref src:_owned) {
+      if isNonNilableClass(this.type) && isNilableClass(src) &&
+         !chpl_legacyNilClasses
+      then
+        compilerError("cannot assign to a non-nilable owned variable from a nilable class");
+
       // Use 'this.type.chpl_t' in case RHS is a subtype
       this.chpl_t = this.type.chpl_t;
       this.chpl_p = src.release();
+      this.complete();
+    }
+
+    proc init=(src: shared) {
+      compilerError("cannot assign to an owned variable from a shared class");
+      this.chpl_t = int; //dummy
```

Comment:
If we're just going to compilerError, are these lines necessary?

Reply:
And if the compiler is yelling about them, that seems like something we should create an issue for

Reply:
Created #13688.